### ISSUE TITLE
tf.keras.metrics.RootMeanSquaredError() not commutative for higher dimensional tensors

### ISSUE THREAD

**Issue by MathiesW** [keras-team/keras#15460]
**System information**.
- Linux Ubuntu 20.04
- Python 3.7.9
- tensorflow-gpu 2.5.0, installed via pip

**Describe the problem, i.e. current behaviour**.
The RootMeanSquaredError metric is not commutative for Tensors with >3 dimensions.

**Describe the expected behavior**.
The RootMeanSquaredError metric should be commutative for all Tensors despite their dimension.

**Standalone code to reproduce the issue**.
```
>>> import tensorflow as tf
>>> import numpy as np

>>> rmse = tf.keras.metrics.RootMeanSquaredError()
>>> y1 = tf.constant(np.random.random((10, 10, 10, 10)))
>>> y2 = tf.constant(np.random.random((10, 10, 10, 10)))

>>> rmse(y1, y2) == rmse(y2, y1)
<tf.Tensor: shape=(), dtype=bool, numpy=False>
```
**Implementation that is commutative**.
```
>>> def rmse_diy(y: tf.Tensor, y_hat: tf.Tensor):
>>>     return tf.sqrt(tf.reduce_mean(tf.square(tf.subtract(y, y_hat))))

>>> rmse_diy(y1, y2) == rmse_diy(y2, y1)
<tf.Tensor: shape=(), dtype=bool, numpy=True>
```

**Comment by sushreebarsa** [comment]
@MathiesW  I tried to run your code on colab using [2.5.0-gpu](https://colab.research.google.com/gist/sushreebarsa/064d13ca878c72688df36c9d872e1699/untitled439.ipynb) ,[2.6.0-gpu](https://colab.research.google.com/gist/sushreebarsa/88360ff0e1d87c157e8e807853ac02db/untitled438.ipynb#scrollTo=F0aPQV4yfg8q) and didn't face the issue reported here .Could you please have a look on the attached gists and let us know if it helps ? Thank you!

**Comment by MathiesW** [comment]
@sushreebarsa thanks for the fast reply. Your attached gists work like a charm. This is interesting, I will have another look at my Python environment tomorrow.

**Comment by LukeWood** [comment]
I have not yet been able to identify the root cause of this unfortunately, but it is definitely a bug.

**Comment by tilakrayal** [comment]
I was able to reproduce the issue on tf-nightly and Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/0871d3b55a4b7f4a4cf10e3182862efd/untitled442.ipynb).

**Comment by ianstenbit** [comment]
Hey @MathiesW -- in your example, it looks like you're not calling `reset_state` on the `rmse` metric in between calls, which is causing it to accumulate previous results in its new result, which is likely what's leading to the small numerical discrepancy.

If you instead use the following code, you'll find the results are identical. The trick is that we need to reset the state of our metric between calls, as Keras metrics are stateful

```
>>> import tensorflow as tf
>>> import numpy as np

>>> rmse = tf.keras.metrics.RootMeanSquaredError()
>>> y1 = tf.constant(np.random.random((10, 10, 10, 10)))
>>> y2 = tf.constant(np.random.random((10, 10, 10, 10)))

>>> r1 = rmse(y1, y2)
>>> rmse.reset_state()
>>> r2 = rmse(y2, y1)
>>> r1 == r2
<tf.Tensor: shape=(), dtype=bool, numpy=True>
```

So I think this is WAI -- please let me know if I've misunderstood your question. Thanks!
